lightningd/onchain/onchain.c

```diff
@@ -0,0 +1,1422 @@
+		    && scripteq(tx->output[i].script, script[LOCAL])) {
+			struct bitcoin_tx *to_us;
+			/* BOLT #5:
+			 *
+			 * 1. _A's main output_: A node SHOULD spend this
+			 *    output to a convenient address.
+			 */
+			out = new_tracked_output(&outs, txid, tx_blockheight,
+						 OUR_UNILATERAL, i,
+						 tx->output[i].amount,
+						 DELAYED_OUTPUT_TO_US);
+			/* BOLT #3:
+			 *
+			 * It is spent by a transaction with `nSequence` field
+			 * set to `to_self_delay` (which can only be valid
+			 * after that duration has passed), and witness:
+			 *
+			 *	<local_delayedsig> 0
+			 */
+			to_us = tx_to_us(out, out, to_self_delay, 0,
+					 feerate_per_kw,
+					 local_dust_limit_satoshi,
+					 local_wscript, our_wallet_pubkey,
+					 &local_delayedprivkey,
+					 &keyset.self_delayed_payment_key);
+
+			/* BOLT #5:
+			 *
+			 * If the output is spent (as recommended), the output
+			 * is *resolved* by the spending transaction */
+			propose_resolution(out, to_us, to_self_delay,
+					   OUR_UNILATERAL_TO_US_RETURN_TO_WALLET);
+
+			script[LOCAL] = NULL;
+			continue;
+		}
+		if (script[REMOTE]
+		    && scripteq(tx->output[i].script, script[REMOTE])) {
+			/* BOLT #5:
+			 *
+			 * 2. _B's main output_: No action required, this
+			 *    output is considered *resolved* by the
+			 *    *commitment transaction* itself. */
+			out = new_tracked_output(&outs, txid, tx_blockheight,
+						 OUR_UNILATERAL, i,
+						 tx->output[i].amount,
+						 OUTPUT_TO_THEM);
+			ignore_output(out);
+			script[REMOTE] = NULL;
+			continue;
+		}
+
+		j = match_htlc_output(tx, i, htlc_scripts);
+		if (j == -1)
+			status_failed(WIRE_ONCHAIN_INTERNAL_ERROR,
+				      "Could not find resolution for output %zu",
+				      i);
+
+		if (htlcs[j].owner == LOCAL) {
+			/* BOLT #5:
+			 *
+			 * 3. _A's offered HTLCs_: See "On-chain HTLC
+			 *    Output Handling: Our Offers" below. */
+			out = new_tracked_output(&outs, txid,
+						 tx_blockheight,
+						 OUR_UNILATERAL, i,
+						 tx->output[i].amount,
+						 OUR_HTLC);
+			resolve_our_htlc_ourcommit(out, htlc_scripts[j],
+						   &htlcs[j], to_self_delay,
+						   &feerate_range,
+						   &local_payment_privkey,
+						   htlc_sigs,
+						   &keyset);
+			/* Each of these consumes one HTLC signature */
+			htlc_sigs++;
+		} else {
+			out = new_tracked_output(&outs, txid,
+						 tx_blockheight,
+						 OUR_UNILATERAL, i,
+						 tx->output[i].amount,
+						 THEIR_HTLC);
+			/* BOLT #5:
+			 *
+			 * 4. _B's offered HTLCs_: See "On-chain HTLC
+			 *    Output Handling: Their Offers" below. */
+			resolve_their_htlc(out, &htlcs[j]);
+		}
+		htlc_scripts[j] = NULL;
+	}
+
+	wait_for_resolved(outs);
+	tal_free(tmpctx);
+}
+
+static void handle_their_cheat(const struct bitcoin_tx *tx,
+			       u64 commit_index,
+			       const struct sha256 *revocation_preimage,
+			       const struct htlc_stub *htlcs,
+			       struct tracked_output **outs)
+{
+	status_failed(WIRE_ONCHAIN_INTERNAL_ERROR,
```

Comment:
This will kill the subdaemon, and the main daemon if we ever encounter a cheat. Sounds a bit harsh, wouldn't a warning be better here? Then again I guess we'll fix this before distributing the client anyway.